### PR TITLE
Add foreign key constraint to `event_forward_extremities`.

### PR DESCRIPTION
**Pull request by erikjohnston**
The bulk of this PR is adding helper functions to make it easier to add constraints. 

This is to help with #13476

Reviewable commit-by-commit.

### CODE REVIEW

**Comment by erikjohnston** [comment]
Note to self: we should probably log when we delete stuff during the validation?

File: synapse/storage/background_updates.py [hunk 1]
```diff
@@ -48,6 +54,78 @@
...
+
+    Used by `register_background_validate_constraint_and_delete_rows`.
+    """
+
+    class State(str, Enum):
+        check = "check"
+        validate = "validate"
+
+    state: State = State.validate
+    lower_bound: Sequence[Any] = ()
```
Comment (DMRobertson) [inline synapse/storage/background_updates.py:126]: Drive-by: are these types being used for validation or is it "just" a dataclass? (If the former, recommend a short unit test to sanity check that the validation behaves as you expect)

Comment (erikjohnston) [reply synapse/storage/background_updates.py:126]: Ermh, it's mostly just used as a data class I think? So long as the pydantic models correctly round trip it should be fine?

File: synapse/storage/background_updates.py [hunk 2]
```diff
@@ -740,6 +818,169 @@ def create_index_sqlite(conn: Connection) -> None:
...
+        update_name: str,
+        table: str,
+        constraint_name: str,
+        constraint: Constraint,
+        unique_columns: Sequence[str],
+    ) -> None:
+        """Helper for store classes to do a background validate constraint, and
+        delete rows that do not pass the constraint check.
+
+        This only applies on PostgreSQL.
```
Comment (MadLittleMods) [inline synapse/storage/background_updates.py:832]: Why? Not applicable to SQLite?

Comment (MadLittleMods) [reply synapse/storage/background_updates.py:832]: It looks like this doesn't apply to SQLite because

> For SQLite the table is recreated and data copied across immediately.

File: synapse/storage/background_updates.py [hunk 3]
```diff
@@ -795,3 +1036,76 @@ def _background_update_progress_txn(
...
+    """Runs a schema delta to add a constraint to the table. This should be run
+    in a schema delta file.
+
+    For PostgreSQL the constraint is added and validated in the background.
+
+    For SQLite the table is recreated and data copied across immediately. This
+    is done by the caller passing in a script to create the new table.
+
+    There must be a corresponding call to
+    `register_background_validate_constraint_and_delete_rows`.
```
Comment (MadLittleMods) [inline synapse/storage/background_updates.py:1061]:
```suggestion
    `register_background_validate_constraint_and_delete_rows` to ensure TODO.
```

File: synapse/storage/background_updates.py [hunk 4]
```diff
@@ -740,6 +818,169 @@ def create_index_sqlite(conn: Connection) -> None:
...
+        self,
+        update_name: str,
+        table: str,
+        constraint_name: str,
+        constraint: Constraint,
+        unique_columns: Sequence[str],
+    ) -> None:
+        """Helper for store classes to do a background validate constraint, and
+        delete rows that do not pass the constraint check.
+
```
Comment (MadLittleMods) [inline synapse/storage/background_updates.py:831]: At first I was confused why we could delete rows without losing data. But we're defining one table to be the source of truth and if there is a row in another table that doesn't have a corresponding row in the source of truth then it's just leftover data that we forgot to delete when we deleted from the source of truth table.

It would be nice to explain the nuance here imo.

File: synapse/storage/background_updates.py [hunk 5]
```diff
@@ -740,6 +818,169 @@ def create_index_sqlite(conn: Connection) -> None:
...
+                await self._background_update_progress(
+                    update_name,
+                    ValidateConstraintProgress(
+                        state=ValidateConstraintProgress.State.check
+                    ).dict(),
+                )
+
+            return batch_size
+        else:
+            raise Exception(f"Unrecognized state '{parsed_progress.state}'")
```
Comment (MadLittleMods) [inline synapse/storage/background_updates.py:982]:
```suggestion
            raise Exception(f"Unrecognized state '{parsed_progress.state}' when trying to validate_constraint_and_delete_in_background")
```

File: synapse/storage/databases/main/event_federation.py [hunk 6]
```diff
@@ -140,6 +141,15 @@ def __init__(
...
         self._clock.looping_call(self._get_stats_for_federation_staging, 30 * 1000)
 
+        if isinstance(self.database_engine, PostgresEngine):
+            self.db_pool.updates.register_background_validate_constraint_and_delete_rows(
+                update_name="event_forward_extremities_constraint_update",
+                table="event_forward_extremities",
+                constraint_name="event_forward_extremities_event_id",
+                constraint=ForeignKeyConstraint("events", [("event_id", "event_id")]),
+                unique_columns=("event_id", "room_id"),
+            )
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/event_federation.py:151]: Why is it most relevant to put the registration in this file vs others? (just curious if any reasoning)

File: synapse/storage/databases/main/events.py [hunk 7]
```diff
@@ -439,6 +433,12 @@ def _persist_events_txn(
 
         self._store_event_txn(txn, events_and_contexts=events_and_contexts)
 
+        self._update_forward_extremities_txn(
+            txn,
+            new_forward_extremities=new_forward_extremities,
+            max_stream_order=max_stream_order,
+        )
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/events.py:440]: Any context behind this order of operations change?

File: synapse/storage/schema/main/delta/78/03event_extremities_constraints.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        room_id TEXT NOT NULL,
+        UNIQUE (event_id, room_id),
+        CONSTRAINT event_forward_extremities_event_id FOREIGN KEY (event_id) REFERENCES events (event_id)
+    );
+"""
+
+FORWARD_EXTREMITIES_INDICES_SCHEMA = """
+    CREATE INDEX ev_extrem_room ON event_forward_extremities(room_id);
+    CREATE INDEX ev_extrem_id ON event_forward_extremities(event_id);
+"""
```
Comment (MadLittleMods) [inline synapse/storage/schema/main/delta/78/03event_extremities_constraints.py:38]: Is there a more automatic way we can copy things instead of relying on our copy-paste skills here to make sure it's right?

File: synapse/storage/schema/main/delta/78/03event_extremities_constraints.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def run_create(cur: LoggingTransaction, database_engine: BaseDatabaseEngine) -> None:
+    run_validate_constraint_and_delete_rows_schema_delta(
+        cur,
+        ordering=7715,
+        update_name="event_forward_extremities_constraint_update",
```
Comment (MadLittleMods) [inline synapse/storage/schema/main/delta/78/03event_extremities_constraints.py:45]: `event_forward_extremities_event_id_foreign_key_constraint_update`

File: synapse/storage/schema/main/delta/78/03event_extremities_constraints.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+FORWARD_EXTREMITIES_INDICES_SCHEMA = """
+    CREATE INDEX ev_extrem_room ON event_forward_extremities(room_id);
+    CREATE INDEX ev_extrem_id ON event_forward_extremities(event_id);
+"""
+
+
+def run_create(cur: LoggingTransaction, database_engine: BaseDatabaseEngine) -> None:
+    run_validate_constraint_and_delete_rows_schema_delta(
+        cur,
+        ordering=7715,
```
Comment (MadLittleMods) [inline synapse/storage/schema/main/delta/78/03event_extremities_constraints.py:44]: How can I check that this `ordering` is reasonable?

File: tests/storage/test_background_update.py [hunk 11]
```diff
@@ -404,3 +411,201 @@ def test_controller(self) -> None:
...
+        self.assertTrue(
+            self.get_success(self.updates.has_completed_background_updates())
+        )
+
+        self.store = self.hs.get_datastores().main
+
+    def test_not_null_constraint(self) -> None:
+        # Create the initial tables, where we have some invalid data.
+        """Tests adding a not null constraint."""
+        sql = """
```
Comment (MadLittleMods) [inline tests/storage/test_background_update.py:431]:
```suggestion
        table_sql = """
```

File: tests/storage/test_background_update.py [hunk 12]
```diff
@@ -404,3 +411,201 @@ def test_controller(self) -> None:
...
+
+        table_sql = """
+            CREATE TABLE test_constraint(
+                a INT PRIMARY KEY,
+                b INT NOT NULL
+            );
+        """
+        self.get_success(
+            self.store.db_pool.execute(
+                "test_not_null_constraint", lambda _: None, base_sql
```
Comment (MadLittleMods) [inline tests/storage/test_background_update.py:532]:
```suggestion
                "test_foreign_key_constraint", lambda _: None, base_sql
```

(applies to other references below)

File: synapse/storage/databases/main/event_federation.py (hunk 6 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/event_federation.py:151]: We just tend to do most operations on `event_forward_extremities` in this file :shrug:

File: synapse/storage/databases/main/events.py (hunk 7 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/events.py:440]: Just that we need to insert the event into the `events` table before we insert into `event_forward_extremities`.

File: synapse/storage/schema/main/delta/78/03event_extremities_constraints.py (hunk 8 above)
Comment (erikjohnston) [reply synapse/storage/schema/main/delta/78/03event_extremities_constraints.py:38]: Actually, probably using the special `sqlite_schema` table. Let me give it a go

File: synapse/storage/schema/main/delta/78/03event_extremities_constraints.py (hunk 10 above)
Comment (erikjohnston) [reply synapse/storage/schema/main/delta/78/03event_extremities_constraints.py:44]: It's meant to match the numbers of the delta, i.e this file is **78**/**03**event_extremities_constraints.py, so this should actually be `7803`

File: synapse/storage/background_updates.py (hunk 4 above)
Comment (erikjohnston) [reply synapse/storage/background_updates.py:831]: I've added a note about double checking that we're OK with deleting rows in `register_background_validate_constraint_and_delete_rows`

File: synapse/storage/schema/main/delta/78/03event_extremities_constraints.py (hunk 8 above)
Comment (erikjohnston) [reply synapse/storage/schema/main/delta/78/03event_extremities_constraints.py]: I have changed it to automatically port the indexes for SQLite, but I can't see a nice way of automating the table creation without having to parse the SQL.

File: synapse/storage/background_updates.py [hunk 13]
```diff
@@ -740,6 +817,177 @@ def create_index_sqlite(conn: Connection) -> None:
...
+        This only applies on PostgreSQL.
+
+        For SQLite the table gets recreated as part of the schema delta.
+
+        Args:
+            update_name: The name of the background update. table: The table
+            with the invalid constraint. constraint_name: The name of the
+            constraint constraint: A `Constraint` object matching the type of
+            constraint. unique_columns: A sequence of columns that form a unique
+            constraint
```
Comment (MadLittleMods) [inline synapse/storage/background_updates.py:844]: Newlines/wrapping messed up here

File: synapse/storage/schema/main/delta/78/03event_extremities_constraints.py (hunk 8 above)
Comment (MadLittleMods) [reply synapse/storage/schema/main/delta/78/03event_extremities_constraints.py]: Fancy!!

---

From some Googling as well, doesn't seem possible for SQLite out of box :+1: 

With Postgres, we could use `CREATE TABLE new_table LIKE old_table;` ([reference](https://www.postgresql.org/docs/current/sql-createtable.html)) which is exactly what we want to copy the schema/structure only.

SQLite does have [`CREATE TABLE ... AS SELECT`](https://www.sqlite.org/lang_createtable.html#create_table_as_select_statements) but it doesn't copy the primary key, constraints, or defaults. It doesn't copy constraints (I assume that means indexes as well) but we could work around it with the code you already have to copy those over. Perhaps we could overcome all of these problems by applying all indexes/constraints (currently the built-in ones are excluded with `AND sql IS NOT NULL`)

https://github.com/matrix-org/synapse/blob/2451d8d61e40109dff11cbe50cf8729c1cc67ad5/synapse/storage/background_updates.py#L1103-L1110

File: synapse/storage/background_updates.py [hunk 14]
```diff
@@ -740,6 +817,177 @@ def create_index_sqlite(conn: Connection) -> None:
...
+        """Helper for store classes to do a background validate constraint, and
+        delete rows that do not pass the constraint check.
+
+        Note: This deletes rows that don't match the constraint. This may not be
+        appropriate in all situations, and so the suitability of using this
+        method should be considered on a case-by-case basis.
+
+        This only applies on PostgreSQL.
+
+        For SQLite the table gets recreated as part of the schema delta.
```
Comment (MadLittleMods) [inline synapse/storage/background_updates.py:837]:
```suggestion
        For SQLite the table gets recreated as part of the schema delta and the data is copied over synchronously (or whatever the correct way to describe it).
```

**Review by MadLittleMods — APPROVED** [review]
(approved in case you want to move on without https://github.com/matrix-org/synapse/pull/15751#discussion_r1237723091, we can iterate on that later)

**Comment by erikjohnston** [comment]
> (approved in case you want to move on without [#15751 (comment)](https://github.com/matrix-org/synapse/pull/15751#discussion_r1237723091), we can iterate on that later)

Thanks! I think its good enough for now, as the biggest footgun was forgetting to port over some indexes, which is harder to do now.
